fix(birthday): updated how the age is calculated

christianfds opened the pull request:
Fixes #2186

christianfds committed:
test(birthday): added failing tests

Signed-off-by: Christian Franchin dos Santos <christianfranchin@gmail.com>

christianfds committed:
fix(birthday): updated how the age is calculated

Uses a simpler way to calculate the age

Signed-off-by: Christian Franchin dos Santos <christianfranchin@gmail.com>

favna commented on tests/lib/Birthday.test.ts:
```diff
@@ -276,6 +276,18 @@ describe('Birthday', () => {
+			});
+
+			test('GIVEN today THEN returns 20', () => {
+				expect(getAge({ year: 2000, month: Month.April, day: 18 })).toBe(20);
```
April 18th 2000 -> April 17th 2020 should be 19
April 18th 2000 -> April 18th 2020 should be 20
April 18th 2000 -> April 19th 2020 should be 20

So the first and last test cases here are incorrect

christianfds replied:
Not actually, if i got it right, what we are changing here is the date of birth, not the current date. So:

Date of birth -> Current Date
April 17th 2000 -> April 18th 2020 should be 20 (birthday was yesterday)
April 18th 2000 -> April 18th 2020 should be 20 (birthday is today)
April 19th 2000 -> April 18th 2020 should be 19 (birthday is tomorrow)

christianfds replied:
I can update the tests descriptions so it becomes more explicity (i also got a bit confused while writting it)

favna replied:
I'm a bit confused right now as well. What I can certainly see is that where we use `getAge` in `upcoming-birthdays` command it does return the correct result right now. 

With the unchanged code it should return 26 for me. That is `February 21st 1995` -> `February 3rd 2022`. Then in `upcoming-birthday` we do `+ 1` on the age and the embed correctly shows 27 for me.

![image](https://user-images.githubusercontent.com/4019718/152416958-edd086c1-7c2d-4a2e-9ce9-45fce799d3d2.png)

Diyagi replied:
https://github.com/skyra-project/skyra/blob/54a95578d7eb5d18b4417c9c173bda5ac318842e/src/commands/Birthday/upcoming-birthdays.ts#L44-L45

So, if im not wrong, when you call `getAge`, you pass my birthdate (mine for exemple, 02/03/1998) and the optional date, which i believe is the date that the next task is schedule (mine for exemple, 02/03/2023).

With the unchanged `getAge`, if you are on the same date as you birthdate, it will still return your old age, so in the exemple above, it would return 24, then you increment +1 (`calculatedAge + 1`) which returns 25.

https://github.com/skyra-project/skyra/blob/82105e6493f005eb342f4a34ab77cc32026e0994/src/tasks/birthday.ts#L99

The problem when the bot announces someone birthday, because we are in the same date as my birthdate, the `getAge` returns my old age, and this value is not incremented like in the `upcoming-birthdays`.

christianfds replied:
I update the method with a different getAge algorithm, if you could test it i would appreciate.

christianfds replied:
I hope that using getUTC... might fix any other issues

christianfds committed:
fix(birthday): fixed ts 2362

It was fixed according to what was pointed out in:

https://github.com/microsoft/TypeScript/issues/5710
Signed-off-by: Christian Franchin dos Santos <christianfranchin@gmail.com>

lveneris commented on src/lib/birthday/utils.ts:
```diff
@@ -111,13 +111,11 @@ export function compareDate(month: Month, day: number, { now = Date.now() }: Tim
-	const years = new Date(now).getUTCFullYear() - data.year;
-	return years + extra;
+	const nowAsDate = new Date(now);
+	const dataAsDate = new Date(data.year, data.month - 1, data.day);
```
These variables are poorly named since both were dates of some kind strictly speaking. Since `nowAsDate` is erroneous (see the following comment), remove it and ideally rename `dataAsDate` to something more fitting.

Alternatively, you could drop the intermediate variable since it adds no readability and drop the data parse into the years calculation.

lveneris commented on src/lib/birthday/utils.ts:
```diff
@@ -111,13 +111,11 @@ export function compareDate(month: Month, day: number, { now = Date.now() }: Tim
+	const nowAsDate = new Date(now);
+	const dataAsDate = new Date(data.year, data.month - 1, data.day);
+
+	const years = new Date(Number(nowAsDate) - Number(dataAsDate)).getFullYear() - new Date(0).getFullYear() - 1;
```
There is additionally no point using `new Date(now)` if you're simply going to convert it back into a number. `Date.now()` returns the same number you'd get from doing `Number(new Date(Date.now()))`, so there's no need to change `now` to a date.

Furthermore, that ternary seems to have been necessary - adding the unconditional -1 returns the wrong age for my birthday following these exact calculations, but I'm not quite sure the best way to fix that. Strangely, in my testing it works perfectly if you remove the -1 from the calculation, but that'd require further verification.

Update: It stands to reason mathematically that even on your birthday, using this standard method, subtracting 1 is never necessary. `now - birthYear = age` is always accurate when the dates are contrasted in milliseconds rather than years, which they are here. Therefore, the -1 need not be there. I verified this on both my birthday, the current date, and Fav's birthday, and all 3 returned correctly without the -1.

```suggestion
	const years = new Date(now - Number(dataAsDate)).getFullYear() - new Date(0).getFullYear();
```

christianfds replied:
I changed it to getUTCFullYear and some issues raised while fixing some others, later i decided to use a "dumb" algorithm.

lveneris replied:
I don't understand why when your initial solution needed a small tweak or two to be perfect.

christianfds committed:
fix(birthday): updated getAge method

Added more tests and gave up on trying to get the age using some smart math

Signed-off-by: Christian Franchin dos Santos <christianfranchin@gmail.com>

lveneris commented on src/lib/birthday/utils.ts:
```diff
@@ -104,20 +104,27 @@ export function compareDate(month: Month, day: number, { now = Date.now() }: Tim
+		age--;
+	}
+
+	return age;
```
```suggestion
	const birthDate = new Date(date.year, date.month - 1, date.day);
	
	return new Date(now - Number(birthDate)).getUTCFullYear() - new Date(0).getUTCFullYear();
```

christianfds committed:
fix(birthday): updated getAge method

co-authored-by: Tyler J Russell <xtylerjrx@gmail.com>

christianfds committed:
fix(birthday): creates date using utc

Signed-off-by: Christian Franchin dos Santos <christianfranchin@gmail.com>

christianfds commented on src/lib/birthday/utils.ts:
```diff
@@ -104,20 +104,16 @@ export function compareDate(month: Month, day: number, { now = Date.now() }: Tim
+export function getAge(date: DateWithOptionalYear, { now = Date.now() }: TimeOptions = {}) {
+	if (date.year === null) return null;
+
+	const birthDate = Date.UTC(date.year, date.month - 1, date.day);
```
Creating it as UTC fixed the issue that was happening.

kyranet commented on src/lib/birthday/utils.ts:
```diff
@@ -104,20 +104,16 @@ export function compareDate(month: Month, day: number, { now = Date.now() }: Tim
+
+	const birthDate = Date.UTC(date.year, date.month - 1, date.day);
+
+	return new Date(now - birthDate).getUTCFullYear() - new Date(0).getUTCFullYear();
```
This is quite the interesting operation, however, we could make it slightly more efficient by not creating a constant over and over. Either create a constant outside this function (and not export), or hardcode `1970` since it's never changing.

Also, it'd be good to same some comment above to show why this works.

As a side note, does this function still work for people born before 1970?

christianfds replied:
Yes, it does work for people born before 1970.

I will add some comments explaining the math of it and probably hardcode 1970

lveneris replied:
> As a side note, does this function still work for people born before 1970?

Yes. Since both `now` and `birthDate` are milliseconds since 1970, `(now - birthDate).getUTCFullYear()` is equivalent to `age + 1970`, so we subtract 1970 and it works.

I'm not sure this needs a comment explaining the maths, honestly. The only thing that may need explaining is the aforementioned offset created by the date system. Hardcoding 1970 would be a good idea, though, since I don't envision the UNIX epoch changing in our lifetimes - alternatively, we could simply convert `(now - birthDate)` to a year, but we may lose precision.

Overall, my recommendation would be to create a top-level constant set to `new Date(0).getUTCFullYear()` as it seems more clear than simply hardcoding 1970 and also avoids the constant being recalculated at every call.

UPDATE: No, the compiler will not sufficiently optimize out the call. My recommendation to make a top-level constant, preferably called something like `EPOCH_YEAR`, still stands.
